### ISSUE TITLE
Declarative Resource Manifests

### ISSUE THREAD

**Issue by sfmskywalker** [OrchardCMS/Orchard#5802]
This would particularly cool for theme developers so that they don't have to write a C# class to implement a resource manifest provider. Instead, all they need is a resource manifest XML file like this:

```XML
<ResourceManifests>
 <Manifest Name="MyResourceManifest" FeatureId="MyFeatureNameToBindTo">
  <Resources>
   <Script Name="MyScript" Url="myscript.min.js" UrlDebug="myscript.js" Dependencies="jQuery" />
   <Style Name="MyStyle" Url="mystyle.min.css" UrlDebug="mystyle.css" />
  </Resources>
 </Manifest>
</ResourceManifests>
```

Themers can then take advantage of `Script.Require` and have dependencies automatically rendered.
The code is already written and is donated by Bede Gaming. It even has some matching API support so that you could for example group resources for a particular device (Mobile or Desktop). For example:

```XML
<ResourceManifests>
	<Manifest MatchProviderType="BedeGaming.DeviceTypeDetectionMatchProvider, BedeGaming">
		<Resources>
			<Script Name="MyCommonScript" Url="MyCommonScript.js" />
			<Group Match="Desktop">
				<Script Name="MyScript" Url="MyScript.Desktop.js" />
			</Group>
			<Group Match="Mobile">
				<Script Name="MyScript" Url="MyScript.Mobile.js" />
			</Group>
		</Resources>
	</Manifest>
</ResourceManifests>
```XML

**Comment by PaulDevenney** [comment]
+1, We're using this currently to separate our our site creating teams from our Orchard engineering team (our themes are in a separate repo to our customised Orchard solution).

**Comment by bleroy** [comment]
More configuration? In a separate file? In XML? Why not put it in the existing manifest, under features?

**Comment by bleroy** [comment]
Also, maybe limit it to simple cases, that don't require inventing a DSL. If you need more complex things (like the groups above), just write it as code, which is better at expressing logic.

**Comment by sfmskywalker** [comment]
Interesting. So that would look something like this:

```
Resources:
    MyScript:
        Type: Script
        Url: MyScript.min.js
        UrlDebug: MyScript.js
    MyStyle:
        Type: Style
        Url: MyStyle.css
        UrlDebug: MyStyle.min.css
```

Or the resource type could even be inferred from the file extension:

```
Resources:
    MyScript:
        Url: MyScript.min.js
        UrlDebug: MyScript.js
    MyStyle:
        Url: MyStyle.css
        UrlDebug: MyStyle.min.css
```

**Comment by bleroy** [comment]
Yup. Looks better, no?

**Comment by sfmskywalker** [comment]
Much easier on the eyes, yes.

However I'm still trying to find a way for them to be able to use this potential enhancement and still be able to match based on something application-specific (in their case, whether or not the request cam from a mobile device). As you said, easiest to just write a custom resource manifest provider class, but their themes don't have .csproj files, so they prefer a declarative approach to resource manifests.

Maybe the class responsible for parsing the Resources section should be replaceable with a custom implementation that understands "groups" and "matching", but then they probably may just as well stick with their current declarative resource manifest provider class that reads XML files.

Despite their case, I think this still seems like a nice addition.

**Comment by bleroy** [comment]
Then let's make that extensible. Then you can build your XML provider with groups and matching as a gallery or private module, and we can keep the shipped implementation simple.

**Comment by sfmskywalker** [comment]
I like it.

Some practical questions regarding implementation:

1. What project should this manifest-reading provider live in. `Orchard.Framework`, a new core module (part of `Orchard.Core`), or a new module project file maybe?
2. Ehm. That's all I have really.

**Comment by bleroy** [comment]
Where's the resource harvesting logic currently? I think it belongs in the same place. Might be some light work to do on settings harvesting to enable it, too.

**Comment by sfmskywalker** [comment]
The resource manager that asks all `IResourceManifestProvider`s lives in Orchard.UI (Framework). Providing resources based on extension manifest files therefore could be done nicely as an implementation of `IResourceManifestProvider`. Framework itself does not currently have any concrete implementations of that interface, so this would be the first one. 

Oh yeah that was another question: should it be bound to a feature or, as you suggest, add a setting.

**Comment by bleroy** [comment]
No, for two reasons: first, that would mean that any feature that wants to take advantage of this would need to take the dependency; second, adding this is non-breaking, and there are no reasons to opt out of it.
A setting would be even worse: if you disable it, then features start breaking. Not good.

**Comment by bleroy** [comment]
Now your third party module's xml provide, now that one will need to be a feature, and modules and themes taking advantage of it will need the dependency.

**Comment by sfmskywalker** [comment]
Ok yeah, that makes sense. 
So then I'm not sure what you mean by "Might be some light work to do on settings harvesting to enable it, too.".

**Comment by bleroy** [comment]
In order to read that additional section under features. Might be that's not necessary, I don't know.

**Comment by sfmskywalker** [comment]
Oh I see. Yeah, actually that reminds me there was some issue somewhere about making manifests extensible. I'll do some digging. Thanks!

**Comment by Jetski5822** [comment]
Side Note: I have made a wrapper around a YAML parser for Brochard... https://github.com/OrchardCMS/Brochard/tree/master/src/Orchard.Parser.Yaml

Makes it nice to parse our settings files and enforces YAML standard. Might be worth porting back?

**Comment by sfmskywalker** [comment]
Interesting. Since I'd have to update the `ExtensionHarvester` class that includes the parsing code anyway, it would make the code much tidier if we could refactor that bit and use a proper YAML parser. I'll check it out.

**Comment by sebastienros** [comment]
Remember we have an `Assets.json` file now. This could be used for this too.

**Comment by Jetski5822** [comment]
I've done shell settings, extension harvester is next on my brochard list.

-----Original Message-----

**Comment by sfmskywalker** [comment]
Interesting. But from the looks of it, the format designed for *Assets.json* is used by some `glob.sync` gulp plugin, so my concern is that putting in anything else will break this gulp process.

**Comment by sfmskywalker** [comment]
I am kind of getting less fond of the idea of mixing resource definitions with either the theme manifest or assets.json file. It seems to me that Theme.txt and Module.txt should not be mixed with resources, but remain focused on theme / module meta data. I don't have any rational explanation for that other than that it feels as if we're violating the SRP. Maybe due to the way I perceive module manifests being responsible for just metadata, and resource manifests are not module metadata. Or is it?

Regarding the concern of 'yet another configuration file', maybe instead of introducing a specific file name, how about introducing a file extension, so that developers can use any file name they like to organize resource manifest files? Could be `Resources.rmx` (*r*esource *m*anifest *x*ml).

These are just thoughts zipping in my head I wanted to get out. Perhaps tomorrow I see things differently yet again.

**Comment by Jetski5822** [comment]
So much configuration... Yaml, xml, json. Hmm

-----Original Message-----

**Comment by bleroy** [comment]
It *is* module metadata, in my opinion.

No new extension, please, no matter what: it breaks too much tooling, and requires extra configuration of the web server in many cases.

**Comment by PaulDevenney** [comment]
The point of the extension (that we need to use in some form regardless of how discussions regarding moving to Orchard core go) is to allow for resouce manifests to be defined outside a module, and specifically within a theme. We have a business trying to create sites at scale. Our site creators do not have access to module content, only the ability to theme it, yet should be allowed to chose their script dependencies etc, and desire to make good use of manifests. From my point of view the key objective is to allow definition inside a theme

It's not like we don't already have "C# resource manifests" right now (so it cannot be argued that this is "more" configuration - the configuration already exists), this is just a nicer option for front-end-only developers who do not wish to (or are not allowed to ) touch the codebase.

Instinctively I would also err on the side of not defining this information in theme.txt, but this is probably because I really don't like the current plain text structure of manifest files (it's hard to externally validate them)

In summary:

* Objective is to allow theme developers more control without access to modules/C#
* We already have resource manifest files, they are just C# currently
* I'm personally not in favour of the info in the current `theme.txt` file, but I can't quite articulate why

It's certainly and interesting discussion :)

**Comment by sebastienros** [comment]
I don't like using the manifests either for that. I'd rather use `Assets.json` or a specific configuration file, but my preference goes to `Assets.json`

**Comment by sebastienros** [comment]
I am opening the bug, but we need to get an agreement on the implementation.
